src/vos/vos_obj.c

```diff
@@ -161,15 +168,21 @@ vos_obj_punch(daos_handle_t coh, daos_unit_oid_t oid, daos_epoch_t epoch,
 		dth->dth_dti_cos_done = 1;
 	}
 
-	/* NB: punch always generate a new incarnation of the object */
+	/* NB: punch always generate a new incarnation of the object.
+	 *
+	 * For non-rebuild case, if the object does not exist, not create it.
+	 */
 	rc = vos_obj_hold(vos_obj_cache_current(), vos_hdl2cont(coh), oid, &epr,
-			  false, DAOS_INTENT_PUNCH, true, &obj);
+			  dth != NULL ? true : false, DAOS_INTENT_PUNCH, true,
```

Comment:
Should we use the REPLAY flag rather than the dth handle to make this decision?

Reply:
You are right. I will fix it.